packages/react-router/modules/Match.js

```diff
@@ -0,0 +1,95 @@
+import React from "react";
+import PropTypes from "prop-types";
+import warning from "tiny-warning";
+import RouterContext from "./RouterContext";
+import matchPath from "./matchPath";
+
+/**
+ * The public API for matching a route and providing that match to children.
+ */
+class Match extends React.Component {
+  render() {
+    return (
+      <RouterContext.Consumer>
+        {context => {
+          if (!context) {
+            throw new Error(
+              __DEV__
+                ? "You should not use <Match> outside a <Router>"
+                : "Invariant failed"
+            );
+          }
+          const location = this.props.location || context.location;
+          let match = this.props.computedMatch
+            ? this.props.computedMatch // <Switch> already computed the match for us
+            : this.props.path
+              ? matchPath(location.pathname, this.props)
+              : context.match;
+
+          const props = { ...context, location, match };
+
+          let { children } = this.props;
+
+          // Preact uses an empty array as children by
+          // default, so use null if that's the case.
+          if (Array.isArray(children) && children.length === 0) {
+            children = null;
+          }
+
+          if (typeof children === "function") {
+            children = children(props);
+
+            if (children === undefined) {
+              if (__DEV__) {
+                const { path } = this.props;
+
+                warning(
+                  false,
+                  "You returned `undefined` from the `children` function of " +
+                    `<Match${path ? ` path="${path}"` : ""}>, but you ` +
+                    "should have returned a React element or `null`"
+                );
+              }
+
+              children = null;
+            }
+          }
+
+          return (
+            <RouterContext.Provider value={props}>
```

Comment:
Just a little improvement: if `children` is `null` we don't need a context provider as wrapper, aren't we? Therefore we can return only `null` and not `null` wrapped inside a context provider.